elyra/contents/parser.py

```diff
@@ -0,0 +1,194 @@
+
+            except KeyError:
+                self.log.warning(f'No language metadata found in {self._filepath}')
+                pass
+
+    @property
+    def language(self) -> str:
+        return self._language
+
+    def read_next_code_chunk(self) -> List[str]:
+        for cell in self._notebook.cells:
+            if cell.source and cell.cell_type == "code":
+                yield cell.source.split('\n')
+
+
+class ScriptParser():
+    """
+    Base class for parsing individual lines of code. Subclasses implement a search_expressions()
+    function that returns language-specific regexes to match against code lines.
+    """
+
+    def parse_environment_variables(self, line):
+        # Parse a line fed from file and match each regex in regex dictionary
+        matches = []
+        for key, value in self.search_expressions().items():
+            for pattern in value:
+                regex = re.compile(pattern)
+                for match in regex.finditer(line):
+                    matches.append((key, match))
+        return matches
+
+
+class PythonScriptParser(ScriptParser):
+    def search_expressions(self) -> Dict[str, List]:
+        # TODO: add more key:list-of-regex pairs to parse for additional resources
+        regex_dict = dict()
+
+        # First regex matches envvar assignments of form os.environ["name"] = value w or w/o value provided
+        # Second regex matches envvar assignments that use os.getenv("name", "value") with ow w/o default provided
+        # Third regex matches envvar assignments that use os.environ.get("name", "value") with or w/o default provided
+        # Both name and value are captured if possible
+        envs = [r"os\.environ\[[\"']([a-zA-Z_]+[A-Za-z0-9_]*)[\"']\](?:\s*=(?:\s*[\"'](.[^\"']*)?[\"'])?)*",
+                r"os\.getenv\([\"']([a-zA-Z_]+[A-Za-z0-9_]*)[\"'](?:\s*\,\s*[\"'](.[^\"']*)?[\"'])?",
+                r"os\.environ\.get\([\"']([a-zA-Z_]+[A-Za-z0-9_]*)[\"'](?:\s*\,(?:\s*[\"'](.[^\"']*)?[\"'])?)*"]
+        regex_dict["env_list"] = envs
+        return regex_dict
+
+
+class RScriptParser(ScriptParser):
+    def search_expressions(self) -> Dict[str, List]:
+        # TODO: add more key:list-of-regex pairs to parse for additional resources
+        regex_dict = dict()
+
+        # Tests for matches of the form Sys.setenv("key" = "value")
+        envs = [r"Sys\.setenv\([\"']*([a-zA-Z_]+[A-Za-z0-9_]*)[\"']*\s*=\s*[\"']*(.[^\"']*)?[\"']*\)",
+                r"Sys\.getenv\([\"']*([a-zA-Z_]+[A-Za-z0-9_]*)[\"']*\)(.)*"]
+        regex_dict["env_list"] = envs
+        return regex_dict
+
+
+class ContentParser(LoggingConfigurable):
+    parsers = {
+        'python': PythonScriptParser(),
+        'r': RScriptParser()
+    }
+
+    def parse(self, filepath: str) -> dict:
+        """Returns a model dictionary of all the regex matches for each key in the regex dictionary"""
+
+        properties = {"env_list": {}, "inputs": {}, "outputs": {}}
```

Comment:
minor: Are we ok with `env_list`? or should this be `env_vars`?

Reply:
Yeah I've thought about this too. I think `env_list` was a holdover from what it was called in the frontend service. Should be a quick change